src/services/game-logic/GameEngine.ts

```diff
@@ -12,16 +12,61 @@ class GameEngine {
 
     getLegalMoves = (square: Square): Square[] => {
         const piece = this.board.getPiece(square);
-        let pieceMoves = piece?.getPossibleMoves(this.board) ?? [];
-        return pieceMoves.filter(this.isOnBoard).filter((square) => !this.isOccupiedBySameColor(square, piece));
+        let legalMoves = piece?.getPossibleMoves(this.board).filter(this.isOnBoard) ?? [];
+        if (piece?.name === PieceNames.KING) {
+            legalMoves = legalMoves.filter((move) => !this.isCastlingIllegal(move, piece));
+        }
+        return legalMoves.filter((destination) => !this.isOccupiedBySameColor(destination, piece));
     };
 
-    private isOnBoard = (square: Square): Boolean => {
+    public movePiece(location: Square, destination: Square): void {
+        //can be useful when saving moves
+        const piece = this.board.getPiece(location);
+        this.board.movePiece(location, destination);
+        if (piece?.name === PieceNames.KING) {
+            this.performCastling(location, destination);
+        }
+    }
+
+    private performCastling(location: Square, destination: Square): void {
+        if (destination.column - location.column === Constants.KINGSIDE_CASTLING) {
+            this.board.movePiece(
+                new Square(location.row, Constants.KINGSIDE_ROOK_COLUMN),
+                new Square(location.row, Constants.KINGSIDE_ROOK_DESTINATION_COLUMN)
+            );
+        } else if (destination.column - location.column === Constants.QUEENSIDE_CASTLING) {
+            this.board.movePiece(
+                new Square(location.row, Constants.QUEENSIDE_ROOK_COLUMN),
+                new Square(location.row, Constants.QUEENSIDE_ROOK_DESTINATION_COLUMN)
+            );
+        }
+    }
+
+    private isCastlingIllegal = (square: Square, piece: Piece): boolean => {
```

Comment:
Tak jak pisał Marcin, raczej unikamy zaprzeczeń w nazwach metod

Reply:
@CaptainObjective  ok, przepisałam tę funkcję i wprowadziłam zmiany, jeśli ok, to poproszę approve :)